### DISCUSSION TITLE
streamdeckc multiaction

### DISCUSSION THREAD

**Discussion by marvin1099** [streamdeck-linux-gui/streamdeck-linux-gui#131] [Ideas / Feature Requests]
I would be great if we can set multiple keys at once to save system resources.  
Then the API connection stays open and completes all actions.  
Maby something like -x and --next.  
When supplied  we split the args list at these places so  
`streamdeckc -a SET_TEXT -p 0 -b 3 -s 0 --text $(date) -x -a SET_TEXT -p 0 -b 1 -s 0 --text $(date)`  
would be the same as  
`streamdeckc -a SET_TEXT -p 0 -b 3 -s 0 --text $(date) ` 
`streamdeckc -a SET_TEXT -p 0 -b 1 -s 0 --text $(date) ` 
but it will keep the API open, therefore it will be faster to run the first command if you want to change multiple keys.  
If you are wondering what wolud this be used for, well let me tell you about my setup.
So i have set up a simple python script to display the time on the streamdeck.  
I'm updating each second since I like seeing seconds  on my steamdeck clock.  
In addition I'm grabbing the my obs recording status and displaying the recoring time on there.  
So if I could i would send the update for the time and the obs recording time as one so a save one command.  
Since I'm doing it each second it would save some CPU cycles.

**Comment by ajbucci** [comment]
I'm pretty sure that the cli server remains running in the background the entire time and listens for incoming socket connections (which should be very lightweight) that are sent when you run streamdeckc. Are you noticing performance issues?

**Reply by marvin1099** [reply]
Jep, there seems to be a bit of a load on my cpu, + - 7% i think.
That seems a bit much for a cli app.
My system is a good gaming system so, that seems strage. 
Im on Vacation rn, so i will do more tests on this when im back.
Also it takes about 0,7 seconds for a cli comand to update the steamdeck, thats why i thoght a websocked is beeing opened and closed

**Reply by ajbucci** [reply]
I could be wrong, but I would think that the image creation to update the button icons every second would contribute much more to the CPU load than any communication overhead.

**Reply by marvin1099** [reply]
Ok but 7% ?
Is there a specific type of image that consumes a lot of resources in order to generate the image ?

**Reply by ajbucci** [reply]
Not sure! The image type shouldn't matter unless it's a gif. Have you tested it with/without the clock to see the difference with your system otherwise at idle? What's the average reported usage when you saturate your cores with something like a benchmark with and without your clock script? Every time you update the time it renders the font as an image with anti aliasing, calculates position on the canvas, and combines the background/foreground layers before sending the final image to the deck.

**Reply by marvin1099** [reply]
I will do more testing, soon.
I will report back then.

**Reply by marvin1099** [reply]
This is that i see in my processes, as you can see streamdeckc uses a bunch on CPU power.
Is this to be expected. Is the streamdeck program itself not supposed use lots of CPU power not streamdeckc.

![clipboard_2024-01-06_11-36](https://github.com/streamdeck-linux-gui/streamdeck-linux-gui/assets/40715852/f5862d43-b9fb-46a8-b95c-cb16c48924d3)

**Reply by marvin1099** [reply]
Any ideas on this, would be great.

**Reply by ajbucci** [reply]
Can you share the script you're using?

**Comment by marvin1099** [comment]
shure thats my script

    #!/home/marvin/.local/pipx/venvs/obsws/bin/python
    import subprocess
    import datetime
    import json
    import time


    def get_obs_status(tout=5):
        tout-=1
        obsrec = False
        obspaused = False
        obstimecode = "00:00:00.000"
        try:
            obsclireturn = subprocess.check_output(["obs-cli", "--password", "%instertobswebpass%", "recording", "status"]).decode("utf-8").strip()
            returnlines = obsclireturn.split('\n')
            for line in returnlines:
                if line.startswith("Recording:"):
                    obsrec = line.split(":")[1].strip().lower() == "true"
                elif line.startswith("Paused:"):
                    obspaused = line.split(":")[1].strip().lower() == "true"
                elif line.startswith("Timecode:"):
                    obstimecode = line.split(":", 1)[1].strip()
        except:
            if tout > -1:
                obstimeli, obsrec, obspaused = get_obs_status(tout)
        obstimeli=obstimecode.split(":")

        return obstimeli, obsrec, obspaused

    def run_ui():
        subprocess.run(["kill", "-9", "streamdeck"]) # this waits
        subprocess.Popen(["streamdeck", "-n"]) # this wont
        time.sleep(0.5)

    def ui_visible():
        try:
            active_window = subprocess.check_output(["xdotool", "search", "--onlyvisible", "--class" ,"StreamDeck UI"]).decode("utf-8").strip()
            if active_window:
                return active_window
            else:
                return False
        except subprocess.CalledProcessError:
            return False

    def fliround(li):
        li[-1] = str(int(float(li[-1]))).rjust(2,'0')
        return ":".join(li)

    def format_timedelta(td):
        seconds = td.seconds
        hours, remainder = divmod(seconds, 3600)
        minutes, seconds = divmod(remainder, 60)
        return f"{hours:02d}:{minutes:02d}:{seconds:02d}"

    if __name__ == "__main__":
        obsrec = False
        obspaused = False
        run_ui()
        subprocess.Popen(["streamdeckc", "-a", "SET_TEXT", "-p", "0", "-b", "7", "-s", "0", "--text", "Start\nRecording"])
        subprocess.Popen(["streamdeckc", "-a", "SET_TEXT", "-p", "0", "-b", "12", "-s", "0", "--text", "^ Press ^\nRecord\nAbove"])
        while True:
            while ui_visible():
                time.sleep(1)
            prev_obsrec = obsrec
            prev_obspaused = obspaused

            current_time = datetime.datetime.now()
            with open('obs.txt') as f:
                cont = "".join(f.readlines()).replace("\n","")
            with open('obs.txt', 'w') as f:
                f.write("")
            if cont == "rec" or cont == "pause":
                lastchecktime = datetime.datetime.now()
                obstimeli, obsrec, obspaused = get_obs_status()
                obsrectimestamp = datetime.timedelta(hours=int(obstimeli[0]), minutes=int(obstimeli[1]), seconds=float(obstimeli[2]))
                #if obsrec and prev_obsrec != obsrec:
                    #obsrectimestamp = datetime.timedelta(hours=int(obstimeli[0]), minutes=int(obstimeli[1]), seconds=float(obstimeli[2]))
                #elif prev_obsrec != obsrec:
                    #obsrectimestamp = datetime.timedelta(hours=int(obstimeli[0]), minutes=int(obstimeli[1]), seconds=float(obstimeli[2]))
                #if obspaused and prev_obspaused != obspaused:
                    #obsrectimestamp = datetime.timedelta(hours=int(obstimeli[0]), minutes=int(obstimeli[1]), seconds=float(obstimeli[2]))
                #elif prev_obspaused != obspaused:
                    #obsrectimestamp = datetime.timedelta(hours=int(obstimeli[0]), minutes=int(obstimeli[1]), seconds=float(obstimeli[2]))
            streamdeck_delay = datetime.timedelta(seconds=0.7)
            ajusttime = current_time + streamdeck_delay
            subprocess.Popen(["streamdeckc", "-a", "SET_TEXT", "-p", "0", "-b", "2", "-s", "0", "--text", ajusttime.strftime('%H:%M:%S\n%d.%m.%Y\n%b %a')])
            if obsrec:
                if obsrec != prev_obsrec:
                    subprocess.run(["streamdeckc", "-a", "SET_TEXT", "-p", "0", "-b", "7", "-s", "0", "--text", "Stop\nRecording"])
                if obspaused and obspaused != prev_obspaused:
                    subprocess.Popen(["streamdeckc", "-a", "SET_TEXT", "-p", "0", "-b", "12", "-s", "0", "--text", "Continue\nRecording\n" + format_timedelta(obsrectimestamp + streamdeck_delay)])
                elif not obspaused:
                    subprocess.Popen(["streamdeckc", "-a", "SET_TEXT", "-p", "0", "-b", "12", "-s", "0", "--text", "Pause\nRecording\n" + format_timedelta(current_time + obsrectimestamp - lastchecktime + streamdeck_delay)])
                # last cheched 0.1 - obs time 0.3 - right now 0.5
                synctime=(lastchecktime.microsecond/1000000 - float(obstimeli[-1]) - datetime.datetime.now().microsecond/1000000 - streamdeck_delay.microseconds/1000000 + 64) % 1
            else:
                if obsrec != prev_obsrec:
                    subprocess.Popen(["streamdeckc", "-a", "SET_TEXT", "-p", "0", "-b", "7", "-s", "0", "--text", "Start\nRecording"])
                    subprocess.Popen(["streamdeckc", "-a", "SET_TEXT", "-p", "0", "-b", "12", "-s", "0", "--text", "^ Press ^\nRecord\nAbove"])
                current_time = datetime.datetime.now()
                synctime = (current_time.microsecond/1000000 - streamdeck_delay.microseconds/1000000 + 1) % 1
            time.sleep(synctime) #sync to obs or time

**Reply by ajbucci** [reply]
Are you using a high resolution image as a button background?

Here are some things I've tried, all examples are digital clocks updating every second using separate calls of streamdeckc:
    - single button using a blank background: <0.5% CPU usage
    - 3 buttons using blank backgrounds: <0.5% CPU usage
    - single button using a very large 10k x 10k pixel image as background: 3-4% CPU usage
    - single button using small image: <0.5% CPU usage

I'm wondering if we're caching the raw image instead of the resized image and resizing every time we draw...

**Reply by marvin1099** [reply]
No Im just using the default blank backround.
I guess that means I might need to do some testing myself.
I realy have no idea why the cpu usage is so bad.

I report back after some more testing

**Reply by ajbucci** [reply]
Thanks! Can you also measure usage with top?

**Comment by marvin1099** [comment]
so i tried this (in bash) for the most simple setup possible (on a black background):

    while true
    do
      streamdeckc -a SET_TEXT -p 0 -b 2 -s 0 --text $(date +"%H:%M:%S")
      sleep $(date +"%3N" | awk '{print (1000-$1)/1000}')
    done
  
And streamdeckc looks like this
![clipboard_2024-01-20_14-22](https://github.com/streamdeck-linux-gui/streamdeck-linux-gui/assets/40715852/57f09b30-4de0-43d1-bf92-35aed5a9f4ae)
Also sometimes 1 or 2 of  the processes seem to use 0.7% (It changes between those 2 a lot)
For me its never seems to be below 0.5% (That could be the cpu / seems it could be a margin of error)

When my python script runns there seem to show 4 processes, so there might be some room for improvements
It seems it wont get better than this.
Thanks for your help but i don't think there is lots of hope for this to be faster.
But you might still want to consider something like -x for sending more than 1 command at once because it seems when i run

    while true
    do 
      streamdeckc -a SET_TEXT -p 0 -b 2 -s 0 --text $(date +"%H:%M:%S") & 
      streamdeckc -a SET_TEXT -p 0 -b 0 -s 0 --text $(date +"%H:%M:%S")
      sleep $(date +"%3N" | awk '{print (1000-$1)/1000}') 
    done

I shows 4 processes and in the worst case all of them use 1.3% so 5.2% in total.
In any case thanks for your inside and help.

**Reply by ajbucci** [reply]
No problem! Just curious, what CPU are you using? Also I'm wondering if it could be an implementation detail between gnome system monitor and kde system monitor. If you measure the CPU usage in top or htop does it match what you see in kde system monitor?

**Reply by marvin1099** [reply]
Take my entire "About this System" tab (cant hurt right):

    Operating System: Garuda Linux 
    KDE Plasma Version: 5.27.10
    KDE Frameworks Version: 5.114.0
    Qt Version: 5.15.12
    Kernel Version: 6.7.0-zen3-1-zen (64-bit)
    Graphics Platform: X11
    Processors: 16 × AMD Ryzen 7 2700X Eight-Core Processor
    Memory: 31,2 GiB of RAM
    Graphics Processor: NVIDIA GeForce GTX 1080 Ti/PCIe/SSE2

**Reply by ajbucci** [reply]
Thanks! K, trying to put my numbers in perspective. I've got a 5600x which is approx ~40% faster in single threaded tasks. So when comparing our numbers my utilization should be roughly ~70% of yours. Also (just being thorough since you probably already know this), since CPU utilization is relative to a single core, your total possible reported utilization is 1600% (8 cores * 2 for SMT * 100%). So when it's reporting 5% that's really ~0.3% of your total potential CPU throughput. 

Having said that, I still think 5% is high... When I did the same test (except in Python), with 3 calls to streamdeckc before the sleep in the while loop, I was getting <0.5% total. This makes me wonder how KDE system monitor is calculating the usage. Can you try using 'top' or 'htop' in the terminal and see if the numbers match what KDE system monitor is reporting?

**Comment by marvin1099** [comment]
So i had a idea like that too after i had run a bunch of cpu usage test yesterday in the evening.
I made shure to avarage the usage out to get an idea of how much performance i realy lose.
I used this script:

    #!/usr/bin/python
    import psutil
    import time
    
    def get_cpu_usage():
        cpu_percent = psutil.cpu_percent(interval=1)
        return cpu_percent
    
    def main():
        mult = 0
        median = 0
        # Track CPU usage for 10 seconds
        for _ in range(30):
            cpu_usage = get_cpu_usage()
            if mult == 0:
                median = cpu_usage
            else:
                median = (median * mult + cpu_usage) / (mult + 1)
    
            mult += 1
    
            time.sleep(1)
            print(f"Current CPU Usage: {cpu_usage}% that results in a average of {median}%")
    
    if __name__ == "__main__":
        main()

And i also ran a test script

    #!/usr/bin/python
    import subprocess
    import threading
    import datetime
    import psutil
    import time
    import os
    
    mult = 0
    median = 0
    cpumed = 0
    while True:
        command = ["streamdeckc", "-a", "SET_TEXT", "-p", "0", "-b", "2", "-s", "0", "--text"]
    
        rn = datetime.datetime.now()
        timestr = rn.strftime('%H:%M:%S\n%d.%m.%Y\n%b %a')
    
        runstr = " ".join(command) + " '" + timestr + "'"
        run = command + [timestr]
        bef = time.time()
        #os.system(runstr)
        subprocess.run(run)
        aft = time.time()
    
        #with lock:
        if mult == 0:
            median = (aft - bef) % 1
        else:
            median = (median * mult + ((aft - bef) % 1)) / (mult + 1)
    
        mult += 1
    
        if (mult % 10) == 0:
            print("On " + str(mult) + " sends it took on average " + str(median))
    
        sleeptime = time.time() % 1
        time.sleep(sleeptime)
    
And here are the results:
My Old script (the one i send first): 9.2% (from 4.8% - 26.2%)
Test script: 5.5% (from 4.0% - 7.8%)
No script: 2.5% (from 1.9% - 4.0%)

So it seems i might be able to somewhat improve on my first script.
Also the system monitor seams like it is capturing moments not averages.
That means your system monitor might be averaging out the usage that your reports seam plausible.
I could also look in something like htop, but i think this is exhaustive enough.

**Reply by marvin1099** [reply]
Hmm i in "htop" rn and if i look at the load average it says:
Old script: 2.44 2.13 1.71
Test script: 2.04 1.47 1.37
No script: 1.56 1.49 1.38

Load average might not be the correct thing to look for.
But if i run "top" i get:
Old script: 7.5us 1.6sy
Test script: 3.6us 0.9sy
No script: 0.5us 0.7sy

These values are changing a lot, but this is kinda the average. 
These also seem to align with the psutil tests.

**Reply by ajbucci** [reply]
Thanks for sharing, these are less concerning, though still a little high perhaps. I think that it probably isn't a big enough issue to dive deeper at the moment though. Thanks again!
